test/dao/InMemoryBookDAO.java

```diff
@@ -0,0 +1,48 @@
+package dao;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+
+import model.Book;
+
+/**
+ * A data-access-object implementation using a purely in-memory "table" for
+ * use in test code.
+ *
+ * @author Jonathan Lovelace
+ */
+public class InMemoryBookDAO implements DataAccessObject<Book> {
+
+	/**
+	 * The table of books.
+	 */
+	private final List<Book> cache = new ArrayList<>();
+
+	@Override
+	public void save(final Book book) {
+		cache.add(book);
+	}
+
+	@Override
+	public void delete(final Book book) {
+		cache.remove(book);
+	}
+
+	@Override
+	public void update(final Book t) {
+		// No-op
```

Comment:
Why do you not have any operation for update?

Reply:
Because in the "in-memory" DAO, the objects themselves are stored in the "database"; if a caller makes a change to an object that has been stored using `save()`, anyone who gets it from `find()` or `findAll()` will automatically see the changes.

Reply:
Oh, I see. That makes sense.